Adjust types to match exported functions

cjbarth opened the pull request:
The `selectWithResolver` type was missing from the type definitions. As I was looking, I noticed a few others things might be wrong. I think this should cover everything. I'll try it in the `xml-crypto` project and report back when that migration is done. Until then, I'd like to get feedback.

JLRishe commented:
Thank you for submitting this.

I'm not all that knowledgeable about .d.ts files and wasn't aware that they could contain functions. Would you mind explaining how one makes use of those functions?

One thing that I notice is that you are using `instanceof Node` and properties on `Node` in these functions, but to the best of my knowledge, the global `Node` object doesn't exist in the node.js environment, only in browsers. Does that mean these functions would fail if someone attempted to use them in node.js? Or does TypeScript take care of that somehow?

cjbarth commented:
Indeed, duck typing does work better for this use-case. in `xml-crypto`, there are more libraries available, so we could get away with what you saw here. I've updated this code to reflect what is actually working in `xml-crypto`. Does this look better to you?

BTW, since the TS refactor of `xml-crypto` depends on this, this PR is a blocker to that getting released.

JLRishe commented on xpath.d.ts:
```diff
@@ -1,11 +1,35 @@
+
+export interface XPathSelect {
+    (expression: string, node: Node): Array<Node>;
+    (expressed: string, node: Node, single: false): Array<Node>;
```
Is there a reason the parameter is called `expressed` here?

cjbarth replied:
Typo. It doesn't affect function though. Fixed.

JLRishe commented on xpath.d.ts:
```diff
@@ -1,11 +1,35 @@
+export type SelectedValue = Node | string | number | boolean | null;
+
+export interface XPathSelect {
+    (expression: string, node: Node): Array<Node>;
```
It looks like these three functions should have the same return types as the three corresponding `select` functions below.

cjbarth replied:
Yes, they do. That is because `useNamespaces` returns a function that follows this interface.

JLRishe replied:
Yes, that's my point. This .d.ts file has them returning different types, but it should have them returning the same types.

cjbarth replied:
I fixed them.

JLRishe commented on xpath.d.ts:
```diff
@@ -1,11 +1,35 @@
-export function useNamespaces(namespaceMap: { [name: string]: string }): XPathSelect;
+
+export type SelectWithResolverReturnType = Array<Node> | SelectedValue;
+export type SelectWithResolverSingleReturnType = SelectedValue;
```
Is there a reason to use this alias instead of using `SelectedValue` directly? Its use on line 15 doesn't really make sense because that `select()` function doesn't involve a resolver.

cjbarth replied:
I've renamed it. I'm not using `SelectedValue` directly because of uniformity. If you'd prefer it to just be `SelectedValue`, I'll gladly make the change.

JLRishe replied:
Ok, I think these names are good.

JLRishe commented on xpath.d.ts:
```diff
@@ -1,11 +1,35 @@
+
+export function useNamespaces(namespaceMap: Record<string, string>): XPathSelect;
+
+export function selectWithResolver(expression: string, node: Node, resolver?: XPathNSResolver | null): SelectWithResolverReturnType;
```
Seems like it would make sense to include an interface definition for `XPathNSResolver` in this file.

I think it would be 

```ts
interface XPathNSResolver {
    lookupNamespaceURI(prefix: string): string;
}
```

cjbarth replied:
Technically, the build-in DOM types already include this. As long as you're using something compatible, we can just use the built-in types.

From `lib.dom.d.ts`
```
type XPathNSResolver = ((prefix: string | null) => string | null) | { lookupNamespaceURI(prefix: string | null): string | null; };
```

JLRishe commented:
Thank you for this update. I've sent you a pull request with some proposed changes and also made some comments.

We should also have unit tests for the new functions you're adding. Can you include those in your PR?

cjbarth commented:
I've made the changes you've discussed.

Side-note: I'd like to format your code, but I don't see a npm script to do so. What formatting tool do you use?

JLRishe commented:
Thank you. I've requested a few more changes to the unit tests, and there's a re-activated comment about the return types in the `XPathSelect` interface.

Until now, I've been careful to leave the whitespace in xpath.js undisturbed so as to not have a huge diff in the history, but I was thinking that it's about time that I added a prettify process. I can do that once this PR is merged.

cjbarth commented:
> I've requested a few more changes to the unit tests

I'm sorry, I don't see these.

Otherwise, I've made all the changes that I think you've requested. Thanks for your review!

cjbarth commented:
> I added a prettify process

IIRC, eslint can format code and it is highly customizable, so you can just pick a style that most closely matches what you have now to minimize changes.

cjbarth commented:
@JLRishe , I can't easily tell from reading the code; can `select(expression: string, node: Node): SelectReturnType;` also be `select(expression: string, node: string): SelectReturnType;`. I'm checking types in `xml-crypto` and I see that we are passing a string to `select()` and it seems to be working. I want to make sure the types are accurate here.

JLRishe commented:
> I'm sorry, I don't see these.

I'm referring to these comments:

![image](https://github.com/goto100/xpath/assets/6964029/5e90b3aa-ca81-466b-81d1-9b09b61ee03d)

> can select(expression: string, node: Node): SelectReturnType; also be select(expression: string, node: string): SelectReturnType;

I don't see any way that would work. 

I think what you're seeing is that `xml-crypto` has a `select(node, xpath)` export that calls off to the `select(xpath, node)` function from the `xpath` package (note the reversal in parameter order):

```js
module.exports.xpath = function (node, xpath) {
  return select(xpath, node);
};
```

And `xml-crypto`'s README.md and example.js are importing and calling that function that has the reversed parameters. Could you double check that and tell me if what you're referring to is somewhere else?

cjbarth commented:
> I'm referring to these comments:

Since they are "Pending", I don't see them. You have to complete the review for them to be visible to me :)

JLRishe commented:
That's weird. I thought I sent them the same way I sent all the previous comments. Hopefully they're visible now.

cjbarth commented:
> Hopefully they're visible now.

They are. I'll resolve these as I push changes related to them. https://github.com/node-saml/xml-crypto/pull/325 is now importing this PR at its HEAD so that I can make sure 100% of these changes are working in real code. I'll convert this to a Draft until it is ready to merge. Thanks again for your attention.

cjbarth commented:
@JLRishe , one of the tests is failing, and it isn't because of any code I wrote. It appears that ProcessingInstruction doesn't have a `nodeName` property. Since it inherits from Node, and Node has a `nodeName`, this is a bug in this library.

```js
    it('should correctly identify a Node of type ProcessingInstruction', () => {
            var doc = parseXml('<book />');
            var pi = doc.createProcessingInstruction('xml-stylesheet', 'href="mycss.css" type="text/css"');

            assert.ok(xpath.isNodeLike(pi));
            assert.ok(xpath.isProcessingInstruction(pi));
            assert.ok(!xpath.isComment(pi));
        });
```

Should have a `nodeName` of "xml-stylesheet", but it is currently `undefined`.

JLRishe commented on xpath.js:
```diff
@@ -4899,25 +4899,30 @@ var xpath = (typeof exports === 'undefined') ? {} : exports;
+    };
+
+    exports.isArrayOfNodes = function (value) {
+        return Array.isArray(value) && value.every(exports.isNodeLike);
```
I think it's a little awkward to be accessing `isNodeLike` as a property of `exports` here and on line 4918.

Could you assign `isNodeLike` and `isArrayOfNodes` to `var`s, include them in the `assign()` call below, and access `isNodeLike` within this file as just `isNodeLike()`?

In the `assign()` call, please remember to include them as `isNodeLike: isNodeLike,` instead of `isNodeLike,`, since we are still maintaining ES5 compatibility for the moment.

cjbarth replied:
I was following the pattern of `select()` and `select1()`:

```js
exports.select1 = function (e, doc) {
    return exports.select(e, doc, true);
};
```

```js
exports.select = function (e, doc, single) {
    return exports.selectWithResolver(e, doc, null, single);
};
```

I'll gladly make whatever change you'd like though. Please confirm you'd like to proceed with your suggestion.

JLRishe replied:
I see, thank you for clarifying.

Yes, please go ahead and make the requested change for `isNodeLike` and `isArrayOfNodes`, but please leave `select1` and `select` as they are.

JLRishe commented:
> It appears that ProcessingInstruction doesn't have a nodeName property. Since it inherits from Node, and Node has a nodeName, this is a bug in this library.

The processing instruction in your test is being created by the `@xmldom/xmldom` package. The problem you're describing has nothing to do with this xpath package.

It does indeed appear that the DOM 3 standard indicates that processing instructions should have a `nodeName` property with the same value as its `target` property, so I would encourage filing an issue on their repo, but I imagine it would be a long time before they addressed it, and having these new functions work with that package is ideal, so please find a suitable workaround. I think removing the `nodeName` check from the `isNodeLike` function would be ok. `(typeof value.nodeName === "string" || typeof value.target === "string")` would also be ok.

cjbarth commented:
> filing an issue on their repo

Done: https://github.com/xmldom/xmldom/issues/505

cjbarth commented:
I've put up a PR to get this fixed in `xmldom`: https://github.com/xmldom/xmldom/pull/509

JLRishe commented on xpath.js:
```diff
@@ -4843,5 +4899,48 @@ var xpath = (typeof exports === 'undefined') ? {} : exports;
+                    && (typeof value.nodeName === "string" || typeof value.target === "string")
+                    && value.nodeType === type;
+            }
+            return exports.isNodeLike(value) && value.nodeType === type;
```
Please just use `isNodeLike()` here.

JLRishe commented on xpath.js:
```diff
@@ -4843,5 +4899,48 @@ var xpath = (typeof exports === 'undefined') ? {} : exports;
+    };
+
+    var isArrayOfNodes = function (value) {
+        return Array.isArray(value) && value.every(isNodeLike);
```
This will not work if any of the nodes are processing instructions.

I think we should just change `isNodeLike` to have `(typeof value.nodeName === "string" || typeof value.target === "string")`. That, along with the checks for `nodeType` should be sufficient to identify a value as nodeLike, and then we can get rid of the special casing in `isNodeOfType`.

cjbarth commented:
This is now no longer needed since `@xmldom/xmldom` pushed a new release. I've reverted the work-arounds and updated the dependency.

FWIW, I note that `engine` in `package.json` specified a minimum version of `0.6`. I tried installing that, and couldn't.

I installed `0.12` and the tests failed because that version of Node doesn't support `const`.

I installed `v4` and that failed because it didn't support destructuring.

I installed `v6` and that failed because `yargs-parser` didn't support anything less than `v10`.

I installed `v10` and the tests passed.

All that to say, the `package-lock.json` was changed to `"lockfileVersion": 1` from `"lockfileVersion": 2` in order to support the older version of Node. I personally recommend doing all development on the oldest-supported version of Node to prevent overlooking these kinds of things.

cjbarth commented:
> Please revert the change to the `"engines"` setting, and the lock file format.

Done. I generated `package-lock.json` using Node@LTS. Since there are no `dependencies`, only `devDependencies`, this shouldn't be a problem for anyone. My apologies for creating extra work for you.



> which has gone far beyond its originally stated purpose

Is there something you'd like to see broken out into multiple PRs?

JLRishe commented:
> Is there something you'd like to see broken out into multiple PRs?

I would have preferred to have the node type test functions in a separate PR. If this were just the .d.ts changes, it could have been merged a lot sooner. But it's ok. Thank you for sticking through this PR.

cjbarth commented:
I'm not worried about the delay as the PR over at `xml-crypto` needs all of this for correct operation. Once this gets released then we can unblock that PR. I appreciate your sticking with me on this mess too; I know it was a lot.

JLRishe commented:
The new version is now published to npm. Thanks!

iahu commented:
that is strange to change the `select` return type to `Array<Node> | SelectedValue`, and this change break my codes.

I think the return type of `select(expression: string, node: Node, single: false)` should be `Array<SelectedValue> | undefined | null`, it should not return a single value type as result.

- if it selected somethings, (if there is just one selected "Value", the result should be transform to `Array<SelectedValue>`)
- if it selected nothing, statement as `null` or `undefined`

BTW, if you make a breaking change, you should public a MAJOR number version.

JLRishe commented:
@iahu I am sorry that this update is causing trouble for you. In this update, the type definitions were modified to be consistent with the actual behavior of the package's implementation. The actual behavior of the package's exported functions has not changed. The type definitions were previously incorrect and have been corrected.

Per the [XPath 1.0 spec](https://www.w3.org/TR/1999/REC-xpath-19991116/), an XPath expression can evaluate to:

- A node set (0 or more nodes)
- A string
- A number
- A boolean

As you can see, it cannot evaluate to a set of strings, or to a null value. This package's `select()` function is consistent with this, so `Array<SelectedValue> | undefined | null` would not be correct.

The author of this PR has added several utility functions that can help to converge the result of `select()` to the correct type, though it looks like there is currently an issue with the `isArrayOfNodes` function: https://github.com/goto100/xpath/pull/121

Personally, I think `select()` isn't very well suited for TypeScript, because of the ambiguity of its return type. For that reason, I strongly suggest using [parsed expressions](https://github.com/goto100/xpath/blob/master/docs/parsed%20expressions.md) which allow evaluating XPaths without such ambiguity.

iahu commented:
after read a bit of the source code. I agree that, the current type declare fit the `select` function's results.

as my above comment, I want get the facts straight. I suggested to change the `select` function's results to be "Array of SelectedValue", which like the `querySelectorAll` DOM API, and as the same time the `select1` just like the `querySelector` API.

cjbarth commented:
> Note: I think there is still a little room for correction in the type definitions since they currently imply that `select(_, _, false)` could return a single `Node` or `null`, which is not true.

I'm reviewing this to make adjustments where needed. Are you hereby saying that `selectWithResolver(_, _, _, false)` will never return Node, `null`, or `undefined`? I think I can see that from the code, but I want to double-check. I'll put up a PR with the other adjustments you've requested so you can comment.
